utils/log_parser.py

```diff
@@ -15,136 +15,224 @@
 
 import argparse
 import re
+from typing import (
+    List,
+    Dict,
+    Tuple,
+)
+import typing
 
 # Maintaining all nodes
-NODES_MAP = {}
-
-# Maintaining all edges
-EDGES_MAP = {}
+NODES_MAP: Dict['NodeNameAndKind', 'Node'] = {}
 
 # Scope stack
-SCOPE_STACK = []
+SCOPE_STACK: List[str] = []
+
+
+class NodeNameAndKind(typing.NamedTuple):
+    name: str
+    kind: str
+
+
+class NodeValue(typing.NamedTuple):
+    node: 'Node'
+    resNo: int
 
 
 class Node:
-    def __init__(self, kindName, name):
-        self.kindName_ = kindName
-        self.name_ = name
-        self.inputs_ = []
-        self.users_ = {}
+    """A class that represents a node in the compute graph
+
+    Public attributes:
+        kindName_: str. The kind name.
+        name_: str. The node name.
+        inputs_: List[NodeValue]. Input node values.
+        users_: Dict['Node', int]. The users of this node.
+    """
+
+    def __init__(self, kindName: str, name: str):
+        self.kindName_: str = kindName
+        self.name_: str = name
+        self.inputs_: List[NodeValue] = []
+        self.users_: Dict['Node', int] = {}
+
+    def get_kind_name(self) -> str:
+        """Gets the kind name. """
 
-    def get_kind_name(self):
         return self.kindName_
 
-    def get_name(self):
+    def get_name(self) -> str:
+        """Gets the node name. """
+
         return self.name_
 
-    def get_inputs(self):
+    def get_inputs(self) -> List[NodeValue]:
+        """Gets the input node. """
+
         return self.inputs_
 
-    def get_users(self):
+    def get_users(self) -> Dict['Node', int]:
+        """Gets the user of this node. """
+
         return self.users_
 
-    def add_user(self, u):
+    def add_user(self, u: 'Node') -> None:
+        """Adds one user of this node. Increment the number of uses of the user by 1. """
+
         if u not in self.users_:
             self.users_[u] = 0
         self.users_[u] += 1
 
-    def remove_user(self, u):
+    def remove_user(self, u: 'Node') -> None:
+        """Removes one use from the given user."""
+
         assert u in self.users_
         self.users_[u] -= 1
         if self.users_[u] == 0:
             del self.users_[u]
 
-    def has_no_uses(self):
-        return len(self.users_) == 0
-
-    def set_input(self, node, resNo):
-        self.inputs_.append((node, resNo))
+    def has_no_uses(self) -> bool:
+        """Returns True if the node has no uses."""
 
-    def replace_input(self, oldNode, oldResNo, newNode, newResNo):
-        assert (oldNode, oldResNo) in self.inputs_
-        tmp_inputs = [(newNode, newResNo)]
-        for n, rn in self.inputs_:
-            if n != oldNode or rn != oldResNo:
-                tmp_inputs.append((n, rn))
-        self.inputs_ = tmp_inputs
+        return len(self.users_) == 0
 
-    def set_scope_of_creation(self, creationScopeName):
+    def set_input(self, nodeVal: NodeValue) -> None:
+        """Adds one input node value."""
```

Comment:
Consider adding a blank line below this line to be consistent with the rest of the formatting.